### ISSUE TITLE
[Bug] div_sfpu / DIV_UNARY_SFPU: tensor÷scalar as x*fl32(1/s) — 1 ULP deviation on ~29% of lanes (fp32 dest)

### ISSUE THREAD

**Issue by Wstedzz** [tenstorrent/tt-metal#54045]
# [Bug] ttnn::div_sfpu / DIV_UNARY_SFPU: tensor÷scalar computed as x * fl32(1/s) — 1 ULP deviation on ~29% of lanes (fp32 dest)

## Summary

The scalar-division path of `binop_with_scalar` (`DIV_UNARY_SFPU`) computes
`x * fl32(1/s)` (double rounding) instead of a correctly-rounded IEEE division.
A source-level model of the kernel (1:1 with the LLK code path) over 4M random
lanes per scalar shows **max 1 ULP deviation from correctly-rounded division on
15–75% of lanes** depending on the scalar (mean 29.2%). The same defect class
was already accepted and fixed for tensor⊗tensor division in #47231 ("Tested
~6.35 Billion pairs, only 74.25% have ULP 0"); this scalar path was left untouched.

| implementation | max rel error | % lanes ≠ RN-exact x/s (s=255 / s=7 / s=3 / mean over 8 scalars) |
|---|---:|---|
| current kernel | 2⁻²³ (1 ULP) | 74.3% / 54.8% / 33.3% / 29.2% |
| proposed fix | 0 | 0% |

bf16 dest masks the defect entirely (0% mismatch); it affects fp32-dest
(`is_fp32_dest_acc_en`) consumers: the public `ttnn::div_sfpu` C++ API,
LLK users of `binop_with_scalar_tile`/`div_unary_tile`, and tt-train kernels
(AdamW, normalization, backward ops) that divide by scalars.

## Root cause

`tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_binop_with_unary.h`
(WH; identical in BH), lines 41–43:

```cpp
} else if constexpr (BINOP_MODE == DIV) {
    // inversion is carried out on host side and passed down
    result = val * parameter;   // parameter = fl32(1/scalar) bitcast from host
```

The host passes `bit_cast<uint32>(1.0f/param)` (`unary_op_utils.cpp:466–469`);
the SFPU performs a single multiply. Double rounding (`1/s` then `x*(1/s)`)
differs from single-rounding `x/s`. Unlike RSUB (which got an explicit
correction comment + fix for logit), this branch has no correction, no
approximation-mode disclaimer, and no documented accuracy caveat.

Note the main user-facing `ttnn.div(tensor, scalar)` overload routes through
binary_ng today, so Python-level regressions are not expected — this affects
the public C++ API surface and direct LLK/Metalium consumers.

## Reproduction (source-level model, no hardware required)

`model_div_scalar.py` (attached below): numpy float32 1:1 model vs
correctly-rounded reference; 4M lanes per scalar across exponent range [2⁻³⁰, 2³¹].

```
s=255 : 74.3% lanes differ (max 1 ULP)
s=7   : 54.8%
s=3   : 33.3%
s=6   : 33.4%   s=10: 20.0%   s=0.001: 16.9%   s=pi: 14.9%   s=0.1: 15.0%
powers of two (e.g. s=0.5): 0% (exact reciprocal)
```

Max relative error observed: 1.192e-07 (= 2⁻²³), consistent with one double-
rounding step.

## Proposed fix

One residual-correction step after the multiply (~4 SFPU instructions), following
the refined-quotient pattern already introduced for binary div in #47231:

```cpp
result = val * parameter;
// refine: q += (x - q*s)*inv   (compensated division)
```

This restores RN-exact `x/s` for all finite inputs at negligible cycle cost.

## Verification plan

Source-level model passes with 0 mismatched lanes post-fix (harness included).
Happy to submit a PR implementing the kernel change + ULP tests once assigned.

---
Related: #47231 (same class, tensor⊗tensor div + RSUB — fixed there, scalar path untouched).

**Comment by beyondpremiere** [comment]
One extension of the same host-side `fl32(1/s)` composition, measured on the Blackhole simulator through the LLK `ScalarDiv` path with the bit pattern ttnn would pass (`std::bit_cast<uint32_t>(1.0f / param0)`, `unary_op_utils.cpp:469`):

For a divisor in the fp32 subnormal range the host reciprocal overflows before the device ever sees it. With `s = 1e-39`, `1.0f / s = +inf`, so `x * inf` is returned for every lane: `+/-inf` on all 968 finite non-zero normal inputs sampled (torch `x / s` is finite for |x| < ~3e-1 of them, e.g. `x = 1.19e-38 -> 11.94`), and `0 / s = 0 * inf = NaN` where torch gives `0`. The mirror case, `s = FLT_MAX`, makes the host reciprocal a subnormal (`2.94e-39`) that the SFPU multiply then flushes; only quotients landing just above the smallest normal are affected there (20 of 1022 lanes), which is the documented SFPMAD denormal-edge behaviour rather than a new defect. Ordinary divisors (3, 7, 0.1, 1e30, 1e-30) reproduce your 1 ULP figure and nothing worse.

Whether a subnormal scalar divisor is inside the op's contract is a judgement call, but the `inf`-everywhere outcome is silent, so noting it here rather than filing separately.
